library/pem.c

```diff
@@ -430,15 +457,13 @@ int mbedtls_pem_read_buffer(mbedtls_pem_context *ctx, const char *header, const
             return ret;
         }
 
-        /*
-         * The result will be ASN.1 starting with a SEQUENCE tag, with 1 to 3
-         * length bytes (allow 4 to be sure) in all known use cases.
-         *
-         * Use that as a heuristic to try to detect password mismatches.
-         */
-        if (len <= 2 || buf[0] != 0x30 || buf[1] > 0x83) {
+        /* Check PKCS padding and update data length based on padding info.
+         * This can be used to detect invalid padding data and password
+         * mismatches. */
+        ret = pem_check_pkcs_padding(buf, len, &len);
+        if (ret != 0) {
             mbedtls_zeroize_and_free(buf, len);
```

Comment:
This will now only zeroize the data and not the padding, since `len` was updated two lines above to exclude the padding. I'd be more comfortable zeroizing the entire buffer.